### ISSUE TITLE
Asynchronous model saving

### ISSUE THREAD

**Issue by gnovis** [Cognexa/cxflow#17]
So far, when the model is being dumped, the whole process of training blocks and waits until the dump is finished which takes ~30s for simple recurrent models. 

It would be nice to dump the model in a separate thread/process in order to speed up this process. However, there are few issues that need to be resolved

1.  Dumping might take more time than finishing one epoch and `BestSaverHook` might conflict with previous instance
2.  The model is mutable. While dumping, it might (and usually is) modified by the training loop. We probably need to create a deep clone of the model. Where is the model stored (GPUmemory/RAM)? How fast is that?

**Comment by blazekadam** [comment]
Removing the critical label as it is not really an issue. At the moment, saving a mediocre model takes a few seconds which happen to be <0.5% of the time spent training.

**Comment by blazekadam** [comment]
This is actually a back-end specific issue. **cxflow** just calls the model `save` method.

Note.: for **cxflow-tensorflow** check out https://stackoverflow.com/questions/41192282/is-there-a-way-to-write-tensorflow-checkpoints-asynchronously
